lab5/DescendentRecursiveParser.cpp

```diff
@@ -0,0 +1,148 @@
+//
+// Created by radoo on 12.12.2022.
+//
+
+#include "DescendentRecursiveParser.h"
+
+void DescendentRecursiveParser::expand(Configuration &config, Grammar grammar,
+                                       const std::vector<std::string> &input) {
+    if (!grammar.isTerminal(config.inputStack.top())) {
+        std::string A = config.inputStack.top();
+        config.inputStack.pop();
+
+        config.workingStack.push(std::make_pair(A, 0));
+
+        auto Ai = grammar.getProductions();
+        auto y1 = Ai[A][0];
+
+        for (int i = y1.getResultList().size() - 1; i >= 0; i--) {
+            config.inputStack.push(y1.getResultList()[i]);
+        }
+    }
+}
+
+void DescendentRecursiveParser::advance(Configuration &config, const Grammar &grammar,
+                                        const std::vector<std::string> &input) {
+    if (config.inputStack.top() == input[config.i]) {
+        std::string terminal = config.inputStack.top();
+        config.inputStack.pop();
+        config.workingStack.push(std::make_pair(terminal, 0));
+        config.i++;
+    }
+}
+
+void DescendentRecursiveParser::momentary_insuccess(Configuration &config, Grammar grammar,
+                                                    const std::vector<std::string> &input) {
+    if (config.inputStack.empty() || grammar.isTerminal(config.inputStack.top()) && config.inputStack.top() != input[config.i]) {
+        config.state = BACK;
+    }
+}
+
+void
+DescendentRecursiveParser::back(Configuration &config, Grammar grammar, const std::vector<std::string> &input) {
```

Comment:
const Grammar and drop 'input'. Also, I think you can pass the grammar by reference. Correct me if I'm wrong, but this is creating a copy of the grammar every time 'back' is invoked